Functions/Quantification/xASL_quant_HadamardDecoding.m

```diff
@@ -107,98 +125,141 @@
-            
-        end
-        idx = idx+xQ.TimeEncodedMatrixSize;
-    end
-end
 
+%% 5. Decode the Hadamard data
+% We assume that the entire encoded volume is reorder into several repetitions of basic Hadamard blocks (repetitions or TE are counted the same)
+decodedPWI4D = zeros(size(imEncodedReordered,1), size(imEncodedReordered,2), size(imEncodedReordered,3), nEncodedVolumes / xQ.TimeEncodedMatrixSize * (xQ.TimeEncodedMatrixSize-1)); 
+decodedBlock = zeros(size(imEncodedReordered,1), size(imEncodedReordered,2), size(imEncodedReordered,3), xQ.TimeEncodedMatrixSize-1); % Temporary single decoded Hadamard block
 
-%% 4. Reorder multi-TE back to the initial order of PLD/TE
-%
-% For model fitting, we want the PLDs-first-TEs-second order (just like the
-% beginning) so we need to reorder it again
+% Go through all repetitions of the Hadamard blocks
+for iRepetition = 1:(nEncodedVolumes / xQ.TimeEncodedMatrixSize)
+	offsetEncoded = xQ.TimeEncodedMatrixSize * (iRepetition-1); % Index offset for the given repetition to search in the encoded data
+	encodedBlock = imEncoded(:,:,:,(1:xQ.TimeEncodedMatrixSize) + offsetEncoded); % Obtain the single encoded block
+	
+	% Decoded the single block
+	for iTI = 1:(xQ.TimeEncodedMatrixSize-1) %Number of decoded PLDs is number of Hadamard matrix size - 1
+		indexPositive = find(xQ.TimeEncodedMatrix(iTI,:)==1);
+		indexNegative = find(xQ.TimeEncodedMatrix(iTI,:)==-1);
+		decodedBlock(:,:,:,iTI) = sum(encodedBlock(:,:,:,(indexPositive)),4) - sum(encodedBlock(:,:,:,(indexNegative)),4);
+	end
+	offsetDecoded = (iRepetition-1)*(xQ.TimeEncodedMatrixSize-1); % Index offset for the given repetition in the decoded data
+	decodedPWI4D(:,:,:,offsetDecoded + (1:(xQ.TimeEncodedMatrixSize-1))) = decodedBlock; % Save the decoded block
+end
+
+%% 6. Reorder multi-TE back to the initial order of PLD/TE
+% For model fitting, we want the PLDs-first-TEs-second order (just like at the beginning) so we need to reorder it again
 if isfield(xQ,'NumberEchoTimes') && (xQ.NumberEchoTimes > 1)
 	% This is the original order
-	vectorOldOrder = 1:size(imDecoded, 4);
+	vectorReorderDecoded = 1:size(decodedPWI4D, 4);
 	
 	% Shape to a matrix with TEs second and all the rest first
-	vectorOldOrder = reshape(vectorOldOrder, size(imDecoded, 4)/xQ.NumberEchoTimes, xQ.NumberEchoTimes);
+	vectorReorderDecoded = reshape(vectorReorderDecoded, size(decodedPWI4D, 4)/xQ.NumberEchoTimes, xQ.NumberEchoTimes);
 	
 	% Flip the two dimensions and make a row vector again
-	vectorOldOrder = reshape(vectorOldOrder', 1, size(imDecoded, 4));
+	vectorReorderDecoded = reshape(vectorReorderDecoded', 1, size(decodedPWI4D, 4));
 
 	% Reorder the data
-	imDecoded = imDecoded(:,:,:,vectorOldOrder);
+	decodedPWI4D = decodedPWI4D(:,:,:,vectorReorderDecoded);
 end
 
-%% 5. Normalization of the decoded data
+% Calculate the correct parameters
```

Comment:
This needs a separate section, and it makes sense perhaps to make this the last section.

Also, is there a way to make the indexDecoding more similar with how you decode the image volumes? E.g., if the "offsetEncoded" is also in indexDecoding but with another name, it would be nice to harmonize these. Now, it is difficult to follow if you do exactly the same for the image and vector decoding.

Reply:
Yes. Of course - a separate section. 

Regarding `offsetEncoded` - I agree with the comment in principle, I'll think of something.

Reply:
Section renaming done in c44ef7a2

Regarding the indexing. This cannot be really done. The Hadamard decoding by summing the individual blocks really follows a completely different logic than work with the parameters. The way we combine and weight the image data has not much in common with the way we reduce the PLD vectors.